File: ml-benchmark/synduce/list/maxhom.f
Inductive List = elt Int | cons {Int, List};
Inductive CList = single Int | concat {CList, CList};

max = \x: Int. \y: Int. if (> x y) then x else y;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt a -> a
  | cons {hd, tl} -> max hd (f tl)
  end
);

cat = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    elt a -> cons {a, ys}
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    single a -> elt a
  | concat {a, b} -> cat (f a) (f b)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/maxhom.ml
type mylist = Elt of int | Cons of int * mylist
type clist = Single of int | Concat of clist * clist

let max x y = if x > y then x else y

let rec spec xs =
  match xs with
  | Elt a -> a
  | Cons (hd, tl) -> max hd (spec tl)

let rec cat xs ys =
  match xs with
  | Elt a -> Cons (a, ys)
  | Cons (a, b) -> Cons (a, cat b ys)

val repr: clist -> mylist compress
let rec repr xs =
  match xs with
  | Single a -> Elt a
  | Concat (a, b) -> cat (repr a) (repr b)

let program x = spec (repr x)